File: SDE_sheet/Arrays/12_count_inversions/solution.go
func numberOfInversions(a []int) int {
    return mergeSort(a, 0, len(a)-1)
}

func mergeSort(arr []int, low, high int) int {
    cnt := 0

    if low >= high {
        return cnt
    }

    mid := low + (high-low)/2

    cnt += mergeSort(arr, low, mid)
    cnt += mergeSort(arr, mid+1, high)
    cnt += merge(arr, low, mid, high)

    return cnt
}

func merge(arr []int, low, mid, high int) int {
    temp := make([]int, 0, high-low+1)
    left := low
    right := mid + 1
    cnt := 0

    for left <= mid && right <= high {
        if arr[left] <= arr[right] {
            temp = append(temp, arr[left])
            left++
        } else {
            temp = append(temp, arr[right])
            cnt += mid - left + 1
            right++
        }
    }

    for left <= mid {
        temp = append(temp, arr[left])
        left++
    }

    for right <= high {
        temp = append(temp, arr[right])
        right++
    }

    for i := low; i <= high; i++ {
        arr[i] = temp[i-low]
    }

    return cnt
}

File: SDE_sheet/Arrays/12_count_inversions/solution.rs
fn number_of_inversions(a: &mut [i32]) -> i32 {
    merge_sort(a, 0, a.len() as isize - 1)
}

fn merge_sort(arr: &mut [i32], low: isize, high: isize) -> i32 {
    let mut cnt = 0;

    if low >= high {
        return cnt;
    }

    let mid = low + (high - low) / 2;

    cnt += merge_sort(arr, low, mid);
    cnt += merge_sort(arr, mid + 1, high);
    cnt += merge(arr, low, mid, high);

    cnt
}

fn merge(arr: &mut [i32], low: isize, mid: isize, high: isize) -> i32 {
    let mut temp = Vec::with_capacity((high - low + 1) as usize);
    let mut left = low;
    let mut right = mid + 1;
    let mut cnt = 0;

    while left <= mid && right <= high {
        if arr[left as usize] <= arr[right as usize] {
            temp.push(arr[left as usize]);
            left += 1;
        } else {
            temp.push(arr[right as usize]);
            cnt += (mid - left + 1) as i32;
            right += 1;
        }
    }

    while left <= mid {
        temp.push(arr[left as usize]);
        left += 1;
    }

    while right <= high {
        temp.push(arr[right as usize]);
        right += 1;
    }

    for i in low..=high {
        arr[i as usize] = temp[(i - low) as usize];
    }

    cnt
}

